### ISSUE TITLE
Brain.NeuralNetworkGPU returns NaN output but works when using Brain.NeuralNetwork

### ISSUE THREAD

**Issue by ashah888** [BrainJS/brain.js#512]
I'm using both Brain.NeuralNetworkGPU and Brain.NeuralNetwork with exactly the same config and data but Brain.NeuralNetworkGPU always returns NaN output value. Also the training for Brain.NeuralNetworkGPU always returns 0 error % and finishes immediately which leads me to believe its not training at all.

**Comment by robertleeplummerjr** [comment]
The template for issues is there to help us help you.  In case you missed it, here it is again.

<img width="638" alt="Screen Shot 2020-02-25 at 6 23 48 AM" src="https://user-images.githubusercontent.com/679099/75243189-68506980-5797-11ea-9d95-1ebec3f1d8f5.png">

**Comment by ashah888** [comment]
## *What* is wrong?
Exact same use Brain.NeuralNetwork() and Brain.NeuralNetworkGPU() with exact same config and data set fails/doesn't work for GPU but it does for normal.

## *Where* does it happen?
When training, GPU version finishes training immediately with 0 error rate which leads me to believe nothing is happening at all.

Normal version finishes after some time, but at least it trains.

Then when doing the .run() command, GPU version always returns NaN but normal version returns correct Float value.

## *How* do we replicate the issue?
Try the GPU version with same config/dataset as normal version.

I tried it with about 500 dataset with each input containing exactly 400 input length (all 0 or 1) and exactly 7 output length (all 0 or 1).

## *How* important is this?
**5 out of 5**
I can't use the GPU version at all.

**Comment by robertleeplummerjr** [comment]
Can you give an example of reproducing this behavior?

**Comment by ashah888** [comment]
Training the Brain.NeuralNetworkGPU() with an array with a single item that looks like this will fail with 0 error threshold, and it will return NaN when using run().

Note. I trained it with a single array item for testing. This works when using NeuralNetwork().

````
const trainingData = [{
    input: [
  0, 0, 0, 0, 0, 1, 0, 0, 1, 1,
  0, 1, 0, 0, 0, 1, 1, 0, 0, 0,
  0, 0, 0, 1, 1, 1, 0, 0, 1, 1,
  1, 1, 1, 1, 0, 1, 0, 1, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
  0, 1, 0, 0, 0, 0, 0, 1, 0, 0,
  1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
  0, 0
],
    output: [
      0, 0, 0, 0, 0, 0, 1
    ]
}]
````

**Comment by ashah888** [comment]
@robertleeplummerjr Can I get a very quick "yes or no" on whether this is a bug or if I'm using it wrong?

**Comment by robertleeplummerjr** [comment]
It looks as though it is working correctly here: https://jsfiddle.net/robertleeplummerjr/nyhab76x/1/

**Comment by ashah888** [comment]
Is there any manual dependencies that could cause it to not work? It literally just works when using NeuralNetwork() but doesn't when I use NeuralNetworkGPU().

**Comment by robertleeplummerjr** [comment]
Are you using this in Node?

**Comment by mubaidr** [comment]
^ Even if the `gpu.js` is not installed, it should fallback to cpu version and work as expected. i.e. `Brain.NeuralNetwork()` and `Brain.NeuralNetworkGPU()` should behave same when gpu is not available.

**Comment by robertleeplummerjr** [comment]
I would double check the size of input you have, I bet it is off.

I just checked in Node, and got this:
```js
const brain = require('./src');

const trainingData = [{
  input: [
    0, 0, 0, 0, 0, 1, 0, 0, 1, 1,
    0, 1, 0, 0, 0, 1, 1, 0, 0, 0,
    0, 0, 0, 1, 1, 1, 0, 0, 1, 1,
    1, 1, 1, 1, 0, 1, 0, 1, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
    0, 1, 0, 0, 0, 0, 0, 1, 0, 0,
    1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
    0, 0
  ],
  output: [
    0, 0, 0, 0, 0, 0, 1
  ]
}];

const net = new brain.NeuralNetworkGPU();
net.train(trainingData);
console.log(net.run([
  0, 0, 0, 0, 0, 1, 0, 0, 1, 1,
  0, 1, 0, 0, 0, 1, 1, 0, 0, 0,
  0, 0, 0, 1, 1, 1, 0, 0, 1, 1,
  1, 1, 1, 1, 0, 1, 0, 1, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
  0, 1, 0, 0, 0, 0, 0, 1, 0, 0,
  1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
  0, 0
]));
/* -> Float32Array [
  0.016257798299193382,
  0.0164403785020113,
  0.016081945970654488,
  0.01629212498664856,
  0.016308937221765518,
  0.016337163746356964,
  0.9834038615226746 ]
*/
```

**Comment by ashah888** [comment]
You see, my problem is even before the .run() command.

When I train using .NeuralNetworkGPU(), it finishes training immediately (even with lots of data) and it *always* finishes with 0 error rate.

I honestly don't think anything is being trained.

**Comment by ashah888** [comment]
@robertleeplummerjr 
I'm sorry but there's definitely some kind of dependency issues or something.

I get [NaN,NaN,NaN,NaN,NaN,NaN,NaN] when I copy and paste exactly the code you put on jsfiddle.

I tried npm rebuild too.

Once again, everything works when I try the same code with normal NeuralNetwork().

I'm using Macbook Pro. In OSX environment. Node JS.

Also tried your code in a brand new Node JS app. There's absolutely no code other than your jsfiddle code.

**Comment by ashah888** [comment]
Exact same code (jsfiddle or my original code) works when I limit the input to 200 array sizes. But once the input is over 200 in size, it fails to train and returns NaN results.

Please help.

**Comment by ErikLarsson82** [comment]
I have the same error, is there anything I can do to debug?

**Comment by ErikLarsson82** [comment]
I found something @ashah888  - over 200 input neurons **and** using GPU needs a config that specifies a large hidden layer.
Node v12.13.1
gpu.js 2.6.2
brain.js - version not sure - a downloaded brain-browser.js file

My input is an array with length 480.
```
new brain.NeuralNetworkGPU({})
//Gives { A: NaN, B: NaN} and {error: 1, iterations: x } 
```

```
new brain.NeuralNetworkGPU({ hiddenLayers: [400] })
//Gives { A: NaN, B: NaN} and {error: 1, iterations: x } 
```

```
new brain.NeuralNetwork({})
//Gives { A: 0.5, B: 0.4} and descending error values for
//each training iteration, this is correct
```
```
new brain.NeuralNetworkGPU({ hiddenLayers: [480] })
//Gives { A: 0.5, B: 0.4} and descending error values for
//each training iteration, this is correct
```

**Comment by ashah888** [comment]
Thank you. I don't understand why these things aren't documented.

**Comment by mubaidr** [comment]
There is no general rule for `hidden layers`. This is very good read here which can help: 
[how-many-hidden-layers-neurons-to-use-in-artificial-neural-networks](https://towardsdatascience.com/beginners-ask-how-many-hidden-layers-neurons-to-use-in-artificial-neural-networks-51466afa0d3e)

**Comment by ErikLarsson82** [comment]
I think you misunderstand @mubaidr. This is simply not a case of a network failing to train properly because the mathematics behind the problem requires more neurons in the hidden layer.

This is _possibly a bug_ where adding (**!**) just one neuron to either the input layer or the hidden layer can be the difference between a **2 second 100% success rate** every time and a network almost ending up at an **error constantly at 1.0** regardless of the amount if training. To my understanding... adding neurons might affect performance somewhat but should not be able to wreck havoc for training.

https://gist.github.com/ErikLarsson82/2078762f7303657af381e6b8b6d3aeca

When i started writing this I thought it was a deterministic effect I observed, however I noticed that when I increase the amount of neurons it's just the probability that increases for the network to get stuck.
